pkg/registry/remote_provider.go

```diff
@@ -0,0 +1,129 @@
+package registry
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/stacklok/toolhive/pkg/networking"
+)
+
+// RemoteRegistryProvider provides registry data from a remote HTTP endpoint
+type RemoteRegistryProvider struct {
+	registryURL  string
+	registry     *Registry
+	registryOnce sync.Once
+	registryErr  error
+}
+
+// NewRemoteRegistryProvider creates a new remote registry provider
+func NewRemoteRegistryProvider(registryURL string) *RemoteRegistryProvider {
+	return &RemoteRegistryProvider{
+		registryURL: registryURL,
+	}
+}
+
+// GetRegistry returns the remote registry data
+func (p *RemoteRegistryProvider) GetRegistry() (*Registry, error) {
+	p.registryOnce.Do(func() {
+		client := networking.GetProtectedHttpClient()
+		resp, err := client.Get(p.registryURL)
+		if err != nil {
+			p.registryErr = fmt.Errorf("failed to fetch registry data from URL %s: %w", p.registryURL, err)
+			return
+		}
+		defer resp.Body.Close()
+
+		// Check if the response status code is OK
+		if resp.StatusCode != http.StatusOK {
+			p.registryErr = fmt.Errorf("response status code from URL %s not OK: status code %d", p.registryURL, resp.StatusCode)
+			return
+		}
+
+		// Read the response body
+		data, err := io.ReadAll(resp.Body)
+		if err != nil {
+			p.registryErr = fmt.Errorf("failed to read registry data from response body: %w", err)
+			return
+		}
+
+		p.registry, p.registryErr = parseRegistryData(data)
+		if p.registryErr != nil {
+			return
+		}
+
+		// Set name field on each server based on map key
+		for name, server := range p.registry.Servers {
+			server.Name = name
+		}
+	})
+
+	return p.registry, p.registryErr
+}
+
+// GetServer returns a specific server by name
+func (p *RemoteRegistryProvider) GetServer(name string) (*Server, error) {
+	reg, err := p.GetRegistry()
+	if err != nil {
+		return nil, err
+	}
+
+	server, ok := reg.Servers[name]
+	if !ok {
+		return nil, fmt.Errorf("server not found: %s", name)
+	}
+
+	return server, nil
+}
+
+// SearchServers searches for servers matching the query
+func (p *RemoteRegistryProvider) SearchServers(query string) ([]*Server, error) {
+	reg, err := p.GetRegistry()
+	if err != nil {
+		return nil, err
+	}
+
+	query = strings.ToLower(query)
+	var results []*Server
+
+	for name, server := range reg.Servers {
```

Comment:
Future consideration: We may want a more efficient way of searching than linear search

Reply:
Oh, definitely 😃